File: Practica3/practica3.py
import time
import random

def max_cruce(A, izq, mid, der):
    suma = 0
    mejor_izq = float('-inf')

    for i in range(mid, izq - 1, -1):
        suma += A[i]
        if suma > mejor_izq:
            mejor_izq = suma

    suma = 0
    mejor_der = float('-inf')

    for i in range(mid + 1, der + 1):
        suma += A[i]
        if suma > mejor_der:
            mejor_der = suma

    return mejor_izq + mejor_der


def max_subarreglo(A, izq, der):
    if izq == der:
        return A[izq]

    mid = (izq + der) // 2

    izquierda = max_subarreglo(A, izq, mid)
    derecha = max_subarreglo(A, mid + 1, der)
    cruce = max_cruce(A, izq, mid, der)

    return max(izquierda, derecha, cruce)


def resolver_archivo():
    try:
        with open("casos.txt", "r", encoding="utf-8") as archivo:
            lineas = [linea.strip() for linea in archivo if linea.strip()]

        t = int(lineas[0])

        for caso in range(1, t + 1):
            datos = list(map(int, lineas[caso].split()))
            n = datos[0]
            A = datos[1:]

            inicio = time.time()
            resultado = max_subarreglo(A, 0, n - 1)
            fin = time.time()

            print(f"\nCaso {caso}:")
            print(A)
            print(f"Suma maxima = {resultado}")
            print(f"Tiempo = {fin - inicio:.6f} segundos")

    except FileNotFoundError:
        print("\nNo se encontro el archivo casos.txt\n")


def pruebas_grandes():
    tamanos = [10, 100, 1000, 10000]

    for n in tamanos:
        A = [random.randint(-100, 100) for _ in range(n)]

        inicio = time.time()
        resultado = max_subarreglo(A, 0, n - 1)
        fin = time.time()

        print(f"\nn = {n}")
        print(f"Suma maxima = {resultado}")
        print(f"Tiempo = {fin - inicio:.6f} segundos")


def menu():
    while True:
        print("\n===== MENU =====")
        print("1. Leer casos.txt")
        print("2. Pruebas grandes")
        print("3. Salir")

        op = input("Opcion: ")

        if op == "1":
            resolver_archivo()

        elif op == "2":
            pruebas_grandes()

        elif op == "3":
            print("Programa finalizado.")
            break

        else:
            print("Opcion invalida.")


menu()

File: Practica3/practica3.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <time.h>

int maximo(int a, int b) {
    return (a > b) ? a : b;
}

int max3(int a, int b, int c) {
    return maximo(maximo(a, b), c);
}

int max_cruce(int A[], int izq, int mid, int der) {
    int suma = 0;
    int mejorIzq = INT_MIN;
    int mejorDer = INT_MIN;

    for (int i = mid; i >= izq; i--) {
        suma += A[i];
        if (suma > mejorIzq)
            mejorIzq = suma;
    }

    suma = 0;

    for (int i = mid + 1; i <= der; i++) {
        suma += A[i];
        if (suma > mejorDer)
            mejorDer = suma;
    }

    return mejorIzq + mejorDer;
}

int max_subarreglo(int A[], int izq, int der) {
    if (izq == der)
        return A[izq];

    int mid = (izq + der) / 2;

    return max3(
        max_subarreglo(A, izq, mid),
        max_subarreglo(A, mid + 1, der),
        max_cruce(A, izq, mid, der)
    );
}

void resolverArchivo() {
    FILE *f = fopen("casos.txt", "r");

    if (f == NULL) {
        printf("\nNo se pudo abrir casos.txt\n");
        return;
    }

    int t;
    fscanf(f, "%d", &t);

    for (int caso = 1; caso <= t; caso++) {
        int n;
        fscanf(f, "%d", &n);

        int *A = (int *)malloc(n * sizeof(int));

        for (int i = 0; i < n; i++)
            fscanf(f, "%d", &A[i]);

        clock_t ini = clock();
        int resultado = max_subarreglo(A, 0, n - 1);
        clock_t fin = clock();

        double tiempo = (double)(fin - ini) / CLOCKS_PER_SEC;

        printf("\nCaso %d:\n", caso);
        for (int i = 0; i < n; i++)
            printf("%d ", A[i]);

        printf("\nSuma maxima = %d\n", resultado);
        printf("Tiempo = %.6f segundos\n", tiempo);

        free(A);
    }

    fclose(f);
}

void pruebasGrandes() {
    int tamanos[] = {10, 100, 1000, 10000};

    for (int k = 0; k < 4; k++) {
        int n = tamanos[k];
        int *A = (int *)malloc(n * sizeof(int));

        for (int i = 0; i < n; i++)
            A[i] = rand() % 201 - 100;

        clock_t ini = clock();
        int resultado = max_subarreglo(A, 0, n - 1);
        clock_t fin = clock();

        double tiempo = (double)(fin - ini) / CLOCKS_PER_SEC;

        printf("\nn = %d\n", n);
        printf("Suma maxima = %d\n", resultado);
        printf("Tiempo = %.6f segundos\n", tiempo);

        free(A);
    }
}

int main() {
    int op;

    do {
        printf("\n===== MENU =====\n");
        printf("1. Leer casos.txt\n");
        printf("2. Pruebas grandes\n");
        printf("3. Salir\n");
        printf("Opcion: ");
        scanf("%d", &op);

        if (op == 1)
            resolverArchivo();
        else if (op == 2)
            pruebasGrandes();
        else if (op == 3)
            printf("Programa finalizado.\n");
        else
            printf("Opcion invalida.\n");

    } while (op != 3);

    return 0;
}